File: nutcracker-app-123/remove_bg.py
from PIL import Image
import os

def remove_background(image_path):
    try:
        img = Image.open(image_path)
        img = img.convert("RGBA")
        datas = img.getdata()

        newData = []
        # Get the color of the top-left pixel to assume as background
        bg_color = datas[0]
        
        # Tolerance for background color matching
        tolerance = 10

        for item in datas:
            # Check if pixel is close to background color
            if (abs(item[0] - bg_color[0]) < tolerance and
                abs(item[1] - bg_color[1]) < tolerance and
                abs(item[2] - bg_color[2]) < tolerance):
                newData.append((255, 255, 255, 0))
            else:
                newData.append(item)

        img.putdata(newData)
        img.save(image_path, "PNG")
        print(f"Processed {image_path}")
    except Exception as e:
        print(f"Error processing {image_path}: {e}")

assets_dir = "/home/dustin/dev/nutcracker/nutcracker-app-123/src/assets"
remove_background(os.path.join(assets_dir, "nutcracker_idle.png"))
remove_background(os.path.join(assets_dir, "nutcracker_cracking.png"))

File: nutcracker-app-123/remove_bg.js
import { Jimp } from 'jimp';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const assetsDir = path.join(__dirname, 'src', 'assets');
const images = ['nutcracker_idle.png', 'nutcracker_cracking.png'];

async function removeBackground(imageName) {
    const imagePath = path.join(assetsDir, imageName);
    try {
        const image = await Jimp.read(imagePath);
        const width = image.bitmap.width;
        const height = image.bitmap.height;

        // Get background color from top-left pixel
        const bgColor = image.getPixelColor(0, 0);
        const bgRGBA = {
            r: (bgColor >>> 24) & 0xFF,
            g: (bgColor >>> 16) & 0xFF,
            b: (bgColor >>> 8) & 0xFF,
            a: bgColor & 0xFF
        };

        console.log(`Processing ${imageName}, BG Color:`, bgRGBA);

        const tolerance = 20; // Adjust tolerance as needed

        image.scan(0, 0, width, height, function (x, y, idx) {
            const r = this.bitmap.data[idx + 0];
            const g = this.bitmap.data[idx + 1];
            const b = this.bitmap.data[idx + 2];
            const a = this.bitmap.data[idx + 3];

            if (
                Math.abs(r - bgRGBA.r) <= tolerance &&
                Math.abs(g - bgRGBA.g) <= tolerance &&
                Math.abs(b - bgRGBA.b) <= tolerance
            ) {
                this.bitmap.data[idx + 3] = 0; // Set alpha to 0
            }
        });

        await image.write(imagePath);
        console.log(`Processed ${imageName}`);
    } catch (err) {
        console.error(`Error processing ${imageName}:`, err);
    }
}

async function main() {
    for (const img of images) {
        await removeBackground(img);
    }
}

main();
